File: Bioinformatics Textbook Track/suffix_tree.cpp
#include "suffix_tree.hpp"

_suffix_tree_edge* _suffix_tree_node::get_edge(char c)
{
    if (this->edges.find(c) == this->edges.end())
        return this->default_map_value;
    return this->edges[c];
}

suffix_tree::suffix_tree(const std::string& text)
{
    _text = text;
    infty = _text.size();
    joker_edge = new _suffix_tree_edge(0, 1, nullptr);
    joker = new _suffix_tree_node();
    root = new _suffix_tree_node(joker);
    joker->set_default(joker_edge);
    joker_edge->target_node = root;
    std::ignore = this->build_tree(root, 0, this->infty);
}

suffix_tree::~suffix_tree()
{
    destroy_node(root);
    destroy_node(joker);
    delete joker_edge;
}

_suffix_tree_node* suffix_tree::build_tree(_suffix_tree_node* node, size_t n, size_t size, size_t skip)
{
    while (n < size)
    {
        char c = this->_text[n];
        _suffix_tree_edge* edge = node->get_edge(c);
        if (edge)
        {
            size_t first = edge->start_pos, last = edge->end_pos;
            size_t i = first, n0 = n;
            if (skip > 0)
            {
                size_t can_skip = std::min(skip, last-first);
                i += can_skip;
                n += can_skip;
                skip -= can_skip;
            }

            while ((i < last) && \
                   (n < size) && \
                   (
                       (this->_text[i] == this->_text[n]) ||
                       (edge == this->joker_edge)
                   )) 
            {
                i++;
                n++;
            }

            if (i == last)
            { // go to the next node
                node = edge->target_node;
            } else { // splitting edge
                _suffix_tree_node* middle_node = new _suffix_tree_node();
                middle_node->set_parent(node);
                middle_node->edges[this->_text[i]] = edge;
                node->edges[c] = new _suffix_tree_edge(first, i, middle_node);
                edge->start_pos = i;
                edge->target_node->set_parent(middle_node);
                middle_node->suff_link = this->build_tree(node->suff_link, n0, n, i-first);
                node = middle_node;
            }

        } else {
            // nowhere to go; creating a new leaf
            _suffix_tree_node* new_leaf = new _suffix_tree_node;
            new_leaf->set_parent(node);
            node->edges[c] = new _suffix_tree_edge(n, this->infty, new_leaf);
            node = node->suff_link;
        }
    }
    return node;
}

void suffix_tree::destroy_node(_suffix_tree_node *node)
{
    if (!node) return;
    for (auto [key, edge]: node->edges)
    {
        if (edge->target_node)
            destroy_node(edge->target_node);
        delete edge;
    }
    delete node;
    node = nullptr;
}

void suffix_tree::get_edges(std::ofstream& ost, _suffix_tree_node* node)
{
    if (!node)
        node = this->root;
    for (const auto [key, edge]: node->edges)
    {
        size_t substr_len = edge->length();
        ost.write(this->_text.c_str()+edge->start_pos, substr_len);
        ost << '\n';
        this->get_edges(ost, edge->target_node);
    }
}

void suffix_tree::lrep(std::ofstream& ost)
{
    size_t maxHeight = 0, substrStart = 0;
    this->lrep_traverse(root, 0, maxHeight, substrStart);
    ost.write(this->_text.c_str()+substrStart, maxHeight);
    ost << '\n';
}

void suffix_tree::lrep_traverse(_suffix_tree_node* node,
                                size_t currHeight,
                                size_t& maxHeight,
                                size_t& suffixIndex)
{
    if (!node) return;

    for (const auto [key, edge]: node->edges)
    {
        size_t substr_len = edge->length();

        if (edge->target_node->isTerminal())
        {
            if (maxHeight < currHeight)
            {
                maxHeight = currHeight;
                suffixIndex = this->_text.size() - currHeight - substr_len;
            }
        } else {
            lrep_traverse(edge->target_node, currHeight + substr_len, maxHeight, suffixIndex);
        }
    }
}

File: Bioinformatics-Stronghold/suffix_tree.py
# Ukkonen's algorithm  for O(n) suffix tree construction
# source: greck, https://habr.com/ru/post/681940/

# data structure:
# node =  < suffix_link, parent, edges >
# edges = { first_char => edge }
# edge =  < idxs, target_node >
# idxs = (first, last) # corresponds to label = text[first: last]

from collections import defaultdict
from typing import Optional

class Node:
    def __init__(self,
                 edges: Optional[defaultdict]=None,
                 suff_link: Optional['Node']=None,
                 parent: Optional['Node']=None):
        self.edges: defaultdict[str, Edge] = edges if edges is not None else defaultdict(type(None))
        self.suff_link: Optional['Node'] = suff_link
        self.parent: Optional['Node'] = parent
    
    def __str__(self) -> str:
        if self.parent:
            idxs = [e.idxs for e in self.parent.edges.values() if e.target_node == self][0]
            return f"{self.parent}{self.text[idxs[0]: idxs[1]]}_"
        else:
            return "_"

    def is_terminal(self) -> bool:
        return len(self.edges) == 0

class Edge:
    def __init__(self,
                 idxs: Optional[tuple[int, int]]=None,
                 target_node: Optional[Node]=None):
        self.idxs = idxs
        self.target_node = target_node

class SuffixTree:
    def __init__(self, text):
        self.text = text
        self.joker_edge = Edge(idxs=(0, 1))
        self.joker = Node(edges=defaultdict(lambda: self.joker_edge))
        self.root = Node(suff_link=self.joker)
        self.joker_edge.target_node = self.root
        self.infty = len(self.text)
        self._build_tree(self.root, 0, self.infty)

    def __repr__(self) -> str:
        return f'{self.__class__.__name__}({self.text})'

    def _build_tree(self, node: Node, n: int, infty: int, skip: int = 0):
        while n < infty:
            c = self.text[n]
            edge = node.edges[c]
            if edge is not None:
                first, last = edge.idxs
                i, n0 = first, n
                if skip > 0:
                    can_skip = min(skip, last - first)
                    i += can_skip
                    n += can_skip
                    skip -= can_skip

                while (
                    i < last and n < infty and
                    (self.text[i] == self.text[n] or edge is self.joker_edge)
                ):
                    i += 1
                    n += 1

                if i == last:  # go to the next node
                    node = edge.target_node
                else:  # splitting edge
                    middle_node = Node(parent=node)
                    middle_node.edges[self.text[i]] = edge
                    node.edges[c] = Edge(idxs=(first, i), target_node=middle_node)
                    edge.idxs = (i, edge.idxs[1])
                    edge.target_node.parent = middle_node
                    middle_node.suff_link = self._build_tree(node.suff_link, n0, n, i - first)
                    node = middle_node
            else:  # no way to go; creating new leaf
                new_leaf = Node(parent=node)
                node.edges[c] = Edge(idxs=(n, self.infty), target_node=new_leaf)
                node = node.suff_link
        return node
    
    def get_edges(self, node=None):
        node = node or self.root
        for c, edge in node.edges.items():
            yield self.text[edge.idxs[0]: edge.idxs[1]]
            yield from self.get_edges(edge.target_node)

    def print_tree(self, node=None, indent=0) -> None:
        Node.text = self.text
        node = node or self.root
        space = "    " * indent
        print(f'{space}ID    : {node}')
        print(f'{space}link  : {node.suff_link}')
        print(f'{space}edges : ')

        for c, edge in node.edges.items():
            print(f'{space}  -{c} {edge.idxs}={self.text[edge.idxs[0]: edge.idxs[1]]}:')
            self.print_tree(edge.target_node, indent + 1)

if __name__ == '__main__':
    text = "CCAAGCTGCTAGAGG$CATGCTGGGCTGGCT%"
    tree = SuffixTree(text)
    tree.print_tree()
